[Adjusted to apply to Debian's 2.6.32 by dann frazier <dannf@debian.org>]

commit ae6df5f96a51818d6376da5307d773baeece4014
Author: Kees Cook <kees.cook@canonical.com>
Date:   Thu Oct 7 10:03:48 2010 +0000

    net: clear heap allocation for ETHTOOL_GRXCLSRLALL
    
    Calling ETHTOOL_GRXCLSRLALL with a large rule_cnt will allocate kernel
    heap without clearing it. For the one driver (niu) that implements it,
    it will leave the unused portion of heap unchanged and copy the full
    contents back to userspace.
    
    Signed-off-by: Kees Cook <kees.cook@canonical.com>
    Acked-by: Ben Hutchings <bhutchings@solarflare.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

diff -urpN linux-source-2.6.32.orig/net/core/ethtool.c linux-source-2.6.32/net/core/ethtool.c
--- linux-source-2.6.32.orig/net/core/ethtool.c	2010-11-24 18:46:41.000000000 -0700
+++ linux-source-2.6.32/net/core/ethtool.c	2010-12-01 20:56:34.199553192 -0700
@@ -265,7 +265,7 @@ static int ethtool_get_rxnfc(struct net_
 	if (info.cmd == ETHTOOL_GRXCLSRLALL) {
 		if (info.rule_cnt > 0) {
 			if (info.rule_cnt <= KMALLOC_MAX_SIZE / sizeof(u32))
-				rule_buf = kmalloc(info.rule_cnt * sizeof(u32),
+				rule_buf = kzalloc(info.rule_cnt * sizeof(u32),
 						   GFP_USER);
 			if (!rule_buf)
 				return -ENOMEM;
